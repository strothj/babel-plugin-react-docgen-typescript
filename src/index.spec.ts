import * as path from "path";
import * as babel from "babel-core";
import plugin, { PluginOptions } from "./index";

it("generates docgen for required props", async () => {
  const result = await transformFile("Component.tsx");

  expect(result.code).toMatchSnapshot();
});

it("omits prop from options", async () => {
  const result = await transformFile("Component.tsx", {
    skipPropsWithName: "onClick",
  });

  expect(result.code).toMatchSnapshot();
});

it("omits multiple props from options", async () => {
  const result = await transformFile("Component.tsx", {
    skipPropsWithName: ["label", "onClick"],
  });

  expect(result.code).toMatchSnapshot();
});

it("adds component to docgen collection", async () => {
  const result = await transformFile("Component.tsx", {
    docgenCollectionName: "STORYBOOK_REACT_CLASSES",
  });

  expect(result.code).toMatchSnapshot();
});

it("extracts literal values from string enum and union types", async () => {
  const result = await transformFile("Component.tsx", {
    shouldExtractLiteralValuesFromEnum: true,
  });

  expect(result.code).toMatchSnapshot();
});

it("extracts literal values from multiple types unions", async () => {
  const result = await transformFile("Component.tsx", {
    shouldExtractValuesFromUnion: true,
  });

  expect(result.code).toMatchSnapshot();
});

function transformFile(fixtureFilename: string, options: PluginOptions = {}) {
  const filePath = path.resolve(__dirname, "__fixtures__", fixtureFilename);

  return new Promise<babel.BabelFileResult>((resolve, reject) => {
    babel.transformFile(
      filePath,
      {
        plugins: [[plugin, options]],
      },
      (err, result) => {
        if (err != null) {
          reject(err);
          return;
        }

        resolve(result);
      },
    );
  });
}
