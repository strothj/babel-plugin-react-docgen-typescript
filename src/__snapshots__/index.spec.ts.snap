// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds component to docgen collection 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require(\\"react\\"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

/**
 * This is an awesome looking button for React.
 */
class Button extends React.Component {
  static defaultProps = {
    label: \\"label\\",
    size: \\"default\\",
    disabled: false,
    tabIndex: -1
  };

  render() {
    const _props = this.props,
          label = _props.label,
          onClick = _props.onClick;
    return React.createElement(React.Fragment, null, React.createElement(\\"button\\", {
      onClick: onClick
    }, label));
  }

}

var _default = Button;
exports.default = _default;

try {
  Button.displayName = \\"Button\\";
  Button.__docgenInfo = {
    description: \\"This is an awesome looking button for React.\\",
    displayName: \\"Button\\",
    props: {
      \\"label\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"label\\"
        },
        \\"description\\": \\"Label for button.\\",
        \\"name\\": \\"label\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"string\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"onClick\\": {
        \\"defaultValue\\": null,
        \\"description\\": \\"Triggered when button is clicked.\\",
        \\"name\\": \\"onClick\\",
        \\"required\\": true,
        \\"type\\": {
          \\"name\\": \\"(event: SyntheticEvent<any>) => void\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"optionalColor\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"#fff\\"
        },
        \\"description\\": \\"An optional color to apply to the button.\\",
        \\"name\\": \\"optionalColor\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"string\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"size\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"default\\"
        },
        \\"description\\": \\"Defines the size of the button.\\",
        \\"name\\": \\"size\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"Size\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"tabIndex\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"-1\\"
        },
        \\"description\\": \\"Defines the tabIndex of the button.\\",
        \\"name\\": \\"tabIndex\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"TabIndex\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"disabled\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"false\\"
        },
        \\"description\\": \\"Defines if a button is disabled.\\",
        \\"name\\": \\"disabled\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"boolean\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      }
    }
  };

  if (typeof STORYBOOK_REACT_CLASSES !== \\"undefined\\") {
    STORYBOOK_REACT_CLASSES[\\"src/__fixtures__/Component.tsx#Button\\"] = {
      name: \\"Button\\",
      docgenInfo: Button.__docgenInfo,
      path: \\"src/__fixtures__/Component.tsx#Button\\"
    };
  }
} catch (e) {}"
`;

exports[`extracts literal values from multiple types unions 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require(\\"react\\"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

/**
 * This is an awesome looking button for React.
 */
class Button extends React.Component {
  static defaultProps = {
    label: \\"label\\",
    size: \\"default\\",
    disabled: false,
    tabIndex: -1
  };

  render() {
    const _props = this.props,
          label = _props.label,
          onClick = _props.onClick;
    return React.createElement(React.Fragment, null, React.createElement(\\"button\\", {
      onClick: onClick
    }, label));
  }

}

var _default = Button;
exports.default = _default;

try {
  Button.displayName = \\"Button\\";
  Button.__docgenInfo = {
    description: \\"This is an awesome looking button for React.\\",
    displayName: \\"Button\\",
    props: {
      \\"label\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"label\\"
        },
        \\"description\\": \\"Label for button.\\",
        \\"name\\": \\"label\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"string\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"onClick\\": {
        \\"defaultValue\\": null,
        \\"description\\": \\"Triggered when button is clicked.\\",
        \\"name\\": \\"onClick\\",
        \\"required\\": true,
        \\"type\\": {
          \\"name\\": \\"(event: SyntheticEvent<any>) => void\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"optionalColor\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"#fff\\"
        },
        \\"description\\": \\"An optional color to apply to the button.\\",
        \\"name\\": \\"optionalColor\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"string\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"size\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"default\\"
        },
        \\"description\\": \\"Defines the size of the button.\\",
        \\"name\\": \\"size\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"enum\\",
          \\"raw\\": \\"Size\\",
          \\"value\\": [\\"\\\\\\"default\\\\\\"\\", \\"\\\\\\"small\\\\\\"\\", \\"\\\\\\"large\\\\\\"\\"]
        }
      },
      \\"tabIndex\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"-1\\"
        },
        \\"description\\": \\"Defines the tabIndex of the button.\\",
        \\"name\\": \\"tabIndex\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"enum\\",
          \\"raw\\": \\"TabIndex\\",
          \\"value\\": [\\"0\\", \\"1\\", \\"-1\\", \\"2\\"]
        }
      },
      \\"disabled\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"false\\"
        },
        \\"description\\": \\"Defines if a button is disabled.\\",
        \\"name\\": \\"disabled\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"enum\\",
          \\"raw\\": \\"boolean\\",
          \\"value\\": [\\"false\\", \\"true\\"]
        }
      }
    }
  };
} catch (e) {}"
`;

exports[`extracts literal values from string enum and union types 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require(\\"react\\"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

/**
 * This is an awesome looking button for React.
 */
class Button extends React.Component {
  static defaultProps = {
    label: \\"label\\",
    size: \\"default\\",
    disabled: false,
    tabIndex: -1
  };

  render() {
    const _props = this.props,
          label = _props.label,
          onClick = _props.onClick;
    return React.createElement(React.Fragment, null, React.createElement(\\"button\\", {
      onClick: onClick
    }, label));
  }

}

var _default = Button;
exports.default = _default;

try {
  Button.displayName = \\"Button\\";
  Button.__docgenInfo = {
    description: \\"This is an awesome looking button for React.\\",
    displayName: \\"Button\\",
    props: {
      \\"label\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"label\\"
        },
        \\"description\\": \\"Label for button.\\",
        \\"name\\": \\"label\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"string\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"onClick\\": {
        \\"defaultValue\\": null,
        \\"description\\": \\"Triggered when button is clicked.\\",
        \\"name\\": \\"onClick\\",
        \\"required\\": true,
        \\"type\\": {
          \\"name\\": \\"(event: SyntheticEvent<any>) => void\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"optionalColor\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"#fff\\"
        },
        \\"description\\": \\"An optional color to apply to the button.\\",
        \\"name\\": \\"optionalColor\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"string\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"size\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"default\\"
        },
        \\"description\\": \\"Defines the size of the button.\\",
        \\"name\\": \\"size\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"enum\\",
          \\"raw\\": \\"Size\\",
          \\"value\\": [{
            \\"value\\": \\"\\\\\\"default\\\\\\"\\"
          }, {
            \\"value\\": \\"\\\\\\"small\\\\\\"\\"
          }, {
            \\"value\\": \\"\\\\\\"large\\\\\\"\\"
          }]
        }
      },
      \\"tabIndex\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"-1\\"
        },
        \\"description\\": \\"Defines the tabIndex of the button.\\",
        \\"name\\": \\"tabIndex\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"TabIndex\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"disabled\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"false\\"
        },
        \\"description\\": \\"Defines if a button is disabled.\\",
        \\"name\\": \\"disabled\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"boolean\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      }
    }
  };
} catch (e) {}"
`;

exports[`generates docgen for required props 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require(\\"react\\"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

/**
 * This is an awesome looking button for React.
 */
class Button extends React.Component {
  static defaultProps = {
    label: \\"label\\",
    size: \\"default\\",
    disabled: false,
    tabIndex: -1
  };

  render() {
    const _props = this.props,
          label = _props.label,
          onClick = _props.onClick;
    return React.createElement(React.Fragment, null, React.createElement(\\"button\\", {
      onClick: onClick
    }, label));
  }

}

var _default = Button;
exports.default = _default;

try {
  Button.displayName = \\"Button\\";
  Button.__docgenInfo = {
    description: \\"This is an awesome looking button for React.\\",
    displayName: \\"Button\\",
    props: {
      \\"label\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"label\\"
        },
        \\"description\\": \\"Label for button.\\",
        \\"name\\": \\"label\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"string\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"onClick\\": {
        \\"defaultValue\\": null,
        \\"description\\": \\"Triggered when button is clicked.\\",
        \\"name\\": \\"onClick\\",
        \\"required\\": true,
        \\"type\\": {
          \\"name\\": \\"(event: SyntheticEvent<any>) => void\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"optionalColor\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"#fff\\"
        },
        \\"description\\": \\"An optional color to apply to the button.\\",
        \\"name\\": \\"optionalColor\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"string\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"size\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"default\\"
        },
        \\"description\\": \\"Defines the size of the button.\\",
        \\"name\\": \\"size\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"Size\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"tabIndex\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"-1\\"
        },
        \\"description\\": \\"Defines the tabIndex of the button.\\",
        \\"name\\": \\"tabIndex\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"TabIndex\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"disabled\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"false\\"
        },
        \\"description\\": \\"Defines if a button is disabled.\\",
        \\"name\\": \\"disabled\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"boolean\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      }
    }
  };
} catch (e) {}"
`;

exports[`omits multiple props from options 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require(\\"react\\"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

/**
 * This is an awesome looking button for React.
 */
class Button extends React.Component {
  static defaultProps = {
    label: \\"label\\",
    size: \\"default\\",
    disabled: false,
    tabIndex: -1
  };

  render() {
    const _props = this.props,
          label = _props.label,
          onClick = _props.onClick;
    return React.createElement(React.Fragment, null, React.createElement(\\"button\\", {
      onClick: onClick
    }, label));
  }

}

var _default = Button;
exports.default = _default;

try {
  Button.displayName = \\"Button\\";
  Button.__docgenInfo = {
    description: \\"This is an awesome looking button for React.\\",
    displayName: \\"Button\\",
    props: {
      \\"optionalColor\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"#fff\\"
        },
        \\"description\\": \\"An optional color to apply to the button.\\",
        \\"name\\": \\"optionalColor\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"string\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"size\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"default\\"
        },
        \\"description\\": \\"Defines the size of the button.\\",
        \\"name\\": \\"size\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"Size\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"tabIndex\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"-1\\"
        },
        \\"description\\": \\"Defines the tabIndex of the button.\\",
        \\"name\\": \\"tabIndex\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"TabIndex\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"disabled\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"false\\"
        },
        \\"description\\": \\"Defines if a button is disabled.\\",
        \\"name\\": \\"disabled\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"boolean\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      }
    }
  };
} catch (e) {}"
`;

exports[`omits prop from options 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require(\\"react\\"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

/**
 * This is an awesome looking button for React.
 */
class Button extends React.Component {
  static defaultProps = {
    label: \\"label\\",
    size: \\"default\\",
    disabled: false,
    tabIndex: -1
  };

  render() {
    const _props = this.props,
          label = _props.label,
          onClick = _props.onClick;
    return React.createElement(React.Fragment, null, React.createElement(\\"button\\", {
      onClick: onClick
    }, label));
  }

}

var _default = Button;
exports.default = _default;

try {
  Button.displayName = \\"Button\\";
  Button.__docgenInfo = {
    description: \\"This is an awesome looking button for React.\\",
    displayName: \\"Button\\",
    props: {
      \\"label\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"label\\"
        },
        \\"description\\": \\"Label for button.\\",
        \\"name\\": \\"label\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"string\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"optionalColor\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"#fff\\"
        },
        \\"description\\": \\"An optional color to apply to the button.\\",
        \\"name\\": \\"optionalColor\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"string\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"size\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"default\\"
        },
        \\"description\\": \\"Defines the size of the button.\\",
        \\"name\\": \\"size\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"Size\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"tabIndex\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"-1\\"
        },
        \\"description\\": \\"Defines the tabIndex of the button.\\",
        \\"name\\": \\"tabIndex\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"TabIndex\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      },
      \\"disabled\\": {
        \\"defaultValue\\": {
          \\"value\\": \\"false\\"
        },
        \\"description\\": \\"Defines if a button is disabled.\\",
        \\"name\\": \\"disabled\\",
        \\"required\\": false,
        \\"type\\": {
          \\"name\\": \\"boolean\\",
          \\"raw\\": null,
          \\"value\\": null
        }
      }
    }
  };
} catch (e) {}"
`;
